compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -420,19 +420,42 @@ abstract class GenJSCode extends plugins.PluginComponent
 
         val topLevelExports = genTopLevelExports(sym)
 
-        val needsStaticInitializer =
-          topLevelExports.exists(_.isInstanceOf[js.TopLevelFieldExportDef])
-        val optStaticInitializer =
-          if (!needsStaticInitializer) Nil
-          else genStaticInitializerLoadingModule(sym) :: Nil
+        memberExports ++ exportedConstructorsOrAccessors ++ topLevelExports
+      }
+
+      // Static initializer
+      val optStaticInitializer = {
+        // Initialization of reflection data, if required
+        val reflectInit = {
+          val enableReflectiveInstantiation = {
+            (sym :: sym.ancestors).exists { ancestor =>
+              ancestor.hasAnnotation(EnableReflectiveInstantiationAnnotation)
+            }
+          }
+          if (enableReflectiveInstantiation)
+            genRegisterReflectiveInstantiation(sym)
+          else
+            Nil
+        }
+
+        // Initialization of the module because of field exports
+        val needsStaticModuleInit =
+          exports.exists(_.isInstanceOf[js.TopLevelFieldExportDef])
+        val staticModuleInit =
+          if (!needsStaticModuleInit) Nil
+          else genLoadModule(sym) :: Nil
 
-        (memberExports ++ exportedConstructorsOrAccessors ++ topLevelExports ++
-            optStaticInitializer)
+        if (reflectInit.nonEmpty || staticModuleInit.nonEmpty) {
+          genStaticInitializerWithStats(
+              js.Block(reflectInit ::: staticModuleInit)) :: Nil
```

Comment:
It was an efficiency concern, but probably cargo-cult based. Concatenating first should be better overall.